### PR TITLE
Maintaining website

### PR DESCRIPTION
**Pull request by jakob-fritz**
New PR instead of #458, not with a branch that was rebased earlier (instead of master merged into branch)

Housekeeping for the website of pySDC including:

- Some updates on RST-Files (e.g. change duplicate names of links)
- Added plots for parallelSDC, so that they are included in the website
- Corrected some paths to images
- Building the website will now fail if images are not found and not included, to highlight this issue early when reappearing

### CODE REVIEW

File: docs/contrib/02_continuous_integration.md [hunk 1]
```diff
@@ -150,7 +150,7 @@ This stage allows to checks how much of the `pySDC` code is tested by the previo
 - `pySDC/tutorial`
 
 This analysis is done in parallel to the test each time a pull is done on any branch (main repository or fork).
-You can look at the current coverage report for the master branch [here](https://parallel-in-time.org/pySDC/coverage/index.html) or compare the results with previous builds [here](https://app.codecov.io/gh/Parallel-in-Time/pySDC). Codecov will also comment on any pull request, indicating the change of coverage.
+You can look at the current coverage report for the [master branch](https://parallel-in-time.org/pySDC/coverage/index.html) or compare the results with [previous builds](https://app.codecov.io/gh/Parallel-in-Time/pySDC). Codecov will also comment on any pull request, indicating the change of coverage.
```
Comment (brownbaerchen) [inline docs/contrib/02_continuous_integration.md:153]: This sounds like the links are leading to the master branch or the previous builds to me. The old version was less ambiguous.

File: pySDC/implementations/convergence_controller_classes/adaptive_collocation.py [hunk 2]
```diff
@@ -16,11 +16,11 @@ class AdaptiveCollocation(ConvergenceController):
...
-    ```
-    params = {
-        'num_nodes': [2, 3],
-    }
-    ```
+
+        params = {
+            'num_nodes': [2, 3],
+        }
+
```
Comment (brownbaerchen) [inline pySDC/implementations/convergence_controller_classes/adaptive_collocation.py:23]: The ticks indicate that this is code. Why did you remove them?

File: pySDC/projects/parallelSDC/preconditioner_playground_MPI.py
Comment (brownbaerchen) [inline pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Does this fix the weird stalling issue that we had when upgrading to new versions of scipy etc?

File: pySDC/implementations/convergence_controller_classes/adaptive_collocation.py (hunk 2 above)
Comment (jakob-fritz) [reply pySDC/implementations/convergence_controller_classes/adaptive_collocation.py:23]: Because they were not rendered correctly (check [the website](https://parallel-in-time.org/pySDC/pySDC/implementations.convergence_controller_classes.adaptive_collocation.html#implementations.convergence_controller_classes.adaptive_collocation.AdaptiveCollocation) ). In local deployment it got rendered correctly with this new syntax

File: pySDC/projects/parallelSDC/preconditioner_playground_MPI.py
Comment (jakob-fritz) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Yes, now the CI works again when plotting the images. This is why the whole thing took me so long...

File: pySDC/implementations/convergence_controller_classes/adaptive_collocation.py (hunk 2 above)
Comment (brownbaerchen) [reply pySDC/implementations/convergence_controller_classes/adaptive_collocation.py:23]: @lisawim, what do you usually do in this case? I believe you have quite some experience with writing documentation :D

File: pySDC/projects/parallelSDC/preconditioner_playground_MPI.py
Comment (jakob-fritz) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: And additionally some minor changes and corrections (e.g. using the with syntax to make use of auto-closing of file or updating error-messages)

File: pySDC/implementations/convergence_controller_classes/adaptive_collocation.py (hunk 2 above)
Comment (tlunet) [reply pySDC/implementations/convergence_controller_classes/adaptive_collocation.py:23]: Actually, code blocks in docstring should probably written like this : https://stackoverflow.com/questions/64451966/python-sphinx-how-to-embed-code-into-a-docstring

Or a more "python-like" approach could be used : https://gitlab.com/tlunet/casper/-/blob/master/casper/maths/analytic.py?ref_type=heads#L79 (not sure how it is rendered by sphinx, tho)

File: docs/contrib/02_continuous_integration.md (hunk 1 above)
Comment (jakob-fritz) [reply docs/contrib/02_continuous_integration.md]: Just updated what the link-text says. Should be more specific and less ambiguous now.

File: pySDC/projects/parallelSDC/preconditioner_playground_MPI.py
Comment (brownbaerchen) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Very nice! Thanks a lot for that! :)

Comment (tlunet) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Lol ... not closing the file made the test stall then :sweat_smile: ...

Comment (jakob-fritz) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Unfortunately not. I gave that a try rather early, but this did not solve the stalling. It may just result in some issue sometime or maybe increased memory or such (as the file is kept open).

Also it did not work out to "just" call the plotting-function as a function, as then un-pickling does not because of the namedTuple.

File: pySDC/implementations/convergence_controller_classes/adaptive_collocation.py (hunk 2 above)
Comment (lisawim) [reply pySDC/implementations/convergence_controller_classes/adaptive_collocation.py:23]: What @tlunet said. :D

Comment (jakob-fritz) [reply pySDC/implementations/convergence_controller_classes/adaptive_collocation.py:25]: Thank you @tlunet for the links. I now included the first way in the file. I didn't know about those and "simply" indenting them already makes it keep formatting, but marking it as code is way better.

File: pySDC/projects/parallelSDC/preconditioner_playground_MPI.py
Comment (tlunet) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Oh ... what solved the stalling issue then ? Not sure to see exactly what from your changes

Comment (jakob-fritz) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Yes. I got the plotting out of the MPI-context. There I got plotting to work but only for a single graph; not for all four of them (still unsure why). This is why I added the command line arguments. With this, I can call the simulation with MPI and later the plotting outside of it. But both calls call the file as script and make it load the definition of the named-tuple. Otherwise loading the pkl-file is not possible.

Comment (tlunet) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Hum ... maybe get rid of those pickle stuff for test could be an idea for https://github.com/orgs/Parallel-in-Time/projects/3, to avoid those kind of issues (which I would guess comes from pickling some MPI communicator or something like that). And use some lightweight text approach, like [this](https://github.com/Parallel-in-Time/pySDC/blob/master/pySDC/helpers/testing.py) for instance ... what do you think @brownbaerchen ?

Comment (brownbaerchen) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Sounds to me like all tasks are opening the file at the same time and the file system will not allow it. I have had similar issues before. Only one task writes the file and that works fine. I guess you essentially open the file only on a single task now as well. Did you try to simply `if comm.rank == 0: <open...>`? I don't think there is a general issue with MPI and pickle.

Comment (jakob-fritz) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Well, pickling the data is still done in the MPI-context. un-pickling also worked in the MPI-context. Even plotting one figure worked in the MPI-context. But plotting multiple figures did not. So it does not seem to be an issue with pickling and MPI.

**Comment by brownbaerchen** [comment]
LGTM

File: pySDC/projects/parallelSDC/preconditioner_playground_MPI.py
Comment (jakob-fritz) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: @brownbaerchen , Yes I did. Only rank==0 writes the file and also rank==0 read it for plotting. Now also only one process reads it as only one process runs this code at that point in time anyway.
But I also assumed initially, that it is some issue with reading the file, which did not turn out to be the case, as the issue was in the actual plotting or writing files (did check which part of it).

Comment (brownbaerchen) [reply pySDC/projects/parallelSDC/preconditioner_playground_MPI.py:1]: Is it to do with writing the plotting file? Still mysterious...

**Comment by lisawim** [comment]
Please correct me if I missed something in our conservation in this PR here but there are still some plots missed in the ``parallelSDC`` projects on the [pySDC website](https://parallel-in-time.org/pySDC/projects/parallelSDC.html). :(

**Comment by brownbaerchen** [comment]
It seems to me like the website was not generated after merging because the post processing step is skipped. Is this because the GitLab action failed?

**Comment by jakob-fritz** [comment]
Yes, the postprocessing step is executed only once both pipelines ran successfully. As now the gitlab-pipeline failed because the "devel" partition for the benchmark is not yet up-and-running.

I/We (Robert and I) decided for this (both pipelines are required), as otherwise e.g. the coverage is misleading, as it sometimes includes coverage from tests from HPC and sometimes it doesn't.
